Fix query remarks on Snowflake

fernandobrito opened the pull request:
Closes https://github.com/metabase/metabase/issues/28865

### Description

Where I work at, we are interested in having better observability of how much cost Metabase is causing on our Snowflake billing.

Metabase already adds remarks (extra metadata added to queries) on all sql-jdbc based drivers (which includes the Snowflake driver). Metabase is prepending the remarks to the query statement, but it turns out that Snowflake actually strips comments prepended to queries. Instead, we should append the comment. There are more details and examples on the issue linked above.

Many drivers are overriding the default remark format (see examples on the issue). This PR also adds a specific format to Snowflake (formatted as JSON and with more metadata), inspired by the remark format used by the Redshift driver.

### How to verify

Trigger any query to Snowflake.

### Demo

Before: no remarks can be seen on Snowflake's query history.

After:

![image](https://user-images.githubusercontent.com/196840/230624551-ae00b523-bfde-48a7-ba7c-f2761e849290.png)

### Checklist

- [x] Signed the Individual Contributor License Agreement
- [x] Tests have been added/updated to cover changes in this PR

### Observation

This is my first PR ever on this repository and also my first time using Clojure (which I'm finding pretty hard). Please review it thoroughly and keep it into account when writing your feedback 😁.

<!-- Reviewable:start -->
- - -
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/metabase/metabase/29888)
<!-- Reviewable:end -->

fernandobrito committed:
Fix query remarks on Snowflake

fernandobrito committed:
Merge branch 'master' into 28865-snowflake-remark

fernandobrito commented:
Hi, @paoliniluis! I'm just missing adding unit tests to cover my changes. Since I have no previous experience with Clojure, I struggled a bit, and then had to work on something else. I'm still planning to come back to this PR and add the tests, but I don't know exactly when I will have time for it.

Is it something that you have the knowledge of and the time to help? 😁

tommyh commented:
@fernandobrito any updates on this?  I might try to take a look as well, but also have never done clojure.

paoliniluis commented:
Hi @fernandobrito, any news?

ian-whitestone commented:
@fernandobrito any plans to revisit this? 🙏

fernandobrito committed:
Merge branch 'master' into 28865-snowflake-remark

fernandobrito committed:
Move docstrings out of defmethod

fernandobrito commented:
Working on this has not been prioritized anymore during my working hours, and I'm quite busy outside of work. Hard to say when I will find time to wrap this up.

fernandobrito commented:
That said, I did get a few hours today to work on this during a "lab day" at work. @paoliniluis do you work for Metabase? How can I get approval from a maintainer so this PR can trigger the GitHub Workflows? I couldn't find anything about it in https://www.metabase.com/docs/latest/developers-guide/start.

paoliniluis commented:
@fernandobrito if the PR is ready and tests are there, we can trigger our CI

fernandobrito committed:
Merge branch 'master' into 28865-snowflake-remark

fernandobrito committed:
Add test to Snowflake query remarks

fernandobrito committed:
Merge branch 'master' into 28865-snowflake-remark

fernandobrito commented:
> @fernandobrito if the PR is ready and tests are there, we can trigger our CI

@paoliniluis thank you. After a couple of hours, I was able to reverse-engineer from the CI steps all the commands and environment variables needed to run the tests locally so I could start working on adding new ones.

I tried my best to add a test to verify my new feature, and I based it on how the Redshift driver tests its query remarks. The PR is now ready to be reviewed and have the CI checks triggered.

metamben commented on modules/drivers/snowflake/src/metabase/driver/snowflake.clj:
```diff
@@ -576,6 +580,27 @@
+  [_ {{:keys [executed-by card-id dashboard-id query-hash context]} :info, query-type :type, database-id :database}]
+  (json/generate-string {:client      "Metabase"
+                         :context     context
+                         :queryType   (case (keyword query-type) :query "MBQL" :native "native")
```
Converting to keyword is not really necessary. I suggest adding a default so that it doesn't blow up if an unexpected `query-type` comes in.

fernandobrito replied:
Done, thanks! I had copied and pasted that code from the query remark implementation from another driver, but now that I think about it, I don't see that much value in having a `case` to map `query` to `MBQL` and there's already a lot of difference in the existing query remarks format and content between different drivers, so I don't see any harm in having it different here. I updated the code to just use `query-type` directly. Let me know if you think it is not a good idea.

metamben replied:
I think that's OK, although the translation might be clearer to newcomers, because native queries are also queries. "MBQL" and "native" are definitely disjoint notions.

metamben commented on modules/drivers/snowflake/src/metabase/driver/snowflake.clj:
```diff
@@ -576,6 +580,27 @@
+                         :cardId      card-id
+                         :dashboardId dashboard-id
+                         :databaseId  database-id
+                         :queryHash   (if (instance? (Class/forName "[B") query-hash)
```
```suggestion
                         :queryHash   (if (bytes? query-hash)
```

fernandobrito replied:
Done, thanks!

metamben commented on modules/drivers/snowflake/test/metabase/driver/snowflake_test.clj:
```diff
@@ -26,6 +29,20 @@
+                  sql-jdbc.execute/execute-statement! check-sql-fn]
+      (u/ignore-exceptions
+        (qp/process-query query))
+      @native-query)))
```
Might be simpler just to throw an exception with `sql` as a parameter (see `ex-info`) and catch it.

metamben commented on src/metabase/driver/sql_jdbc/execute.clj:
```diff
@@ -660,13 +660,26 @@
 
+(defmulti inject-remark
+  "Injects the remark into the SQL query text."
+  {:arglists '([driver sql remark])}
```
It should document the version it was added in.

fernandobrito replied:
Done, thanks!

metamben commented on src/metabase/driver/sql_jdbc/execute.clj:
```diff
@@ -660,13 +660,26 @@
   (let [row-thunk (row-thunk driver rs rsmeta)]
     (qp.reducible/reducible-rows row-thunk canceled-chan)))
 
+(defmulti inject-remark
```
New driver methods should be documented in `docs/developers-guide/driver-changelog.md`.

fernandobrito replied:
Done, thanks!

metamben reviewed (changes requested):
Looks good to me, thanks! I have a few small remarks and requested some documentation.

fernandobrito commented:
> Looks good to me, thanks! I have a few small remarks and requested some documentation.

Thank you for the review, @metamben 🙌. All the comments make sense to me and I will address them as soon as I have time (probably later in the week or in the weekend).

fernandobrito committed:
Fix query remarks on Snowflake

fernandobrito committed:
Merge branch '28865-snowflake-remark' of github.com:fernandobrito/metabase into 28865-snowflake-remark

fernandobrito committed:
Merge branch 'master' into 28865-snowflake-remark

fernandobrito committed:
Merge branch '28865-snowflake-remark' of github.com:fernandobrito/metabase into 28865-snowflake-remark

fernandobrito commented:
I finally had time to try to address your comments, @metamben. Thank you again for the review.

I believe I solved most of the comments, except the suggestion to refactor one of the test functions to use `ex-info`. I do understand the idea behind it, but I did not manage to implement it. This is my first time coding in Clojure and I just copied and pasted the original function from https://github.com/metabase/metabase/blob/master/modules/drivers/redshift/test/metabase/driver/redshift_test.clj#L65.

The closest I got with the help from ChatGPT was:

```
(defn- query->native [query]
  (let [check-sql-fn (fn [_ _ sql & _]
                        (throw (ex-info "done" {:sql sql})))]
    (with-redefs [sql-jdbc.execute/prepared-statement check-sql-fn
                  sql-jdbc.execute/execute-statement! check-sql-fn]
      (try
        (qp/process-query query)
        (catch Exception e
          (when (= (:message e) "done") ; Only capture our exception
            (:sql (ex-data e))))))))
```

But that doesn't work (I get a `java.lang.NullPointerException`). I think the fact that we want to override the implementation of 2 methods but still let them both run is making me struggle with the execution flow. Any suggestions on how I could approach it?

fernandobrito committed:
Fix query remarks on Snowflake

metamben commented:
> But that doesn't work (I get a `java.lang.NullPointerException`). I think the fact that we want to override the implementation of 2 methods but still let them both run is making me struggle with the execution flow. Any suggestions on how I could approach it?

Yes, it's a bit tricky, because the exception thrown is at the end of exception cause chain. I think something like this could work:

```clojure
(defn- query->native [query]
  (let [check-sql-fn (fn [_ _ sql & _]
                       (throw (ex-info "done" {::native-query sql})))]
    (with-redefs [sql-jdbc.execute/prepared-statement check-sql-fn
                  sql-jdbc.execute/execute-statement! check-sql-fn]
      (try
        (qp/process-query query)
        (is false "no statement created")
        (catch Exception e
          (-> e u/all-ex-data ::native-query))))))
```

If it doesn't, don't sweat it, it's not essential.

metamben commented on modules/drivers/snowflake/src/metabase/driver/snowflake.clj:
```diff
@@ -587,6 +591,28 @@
+                         :cardId      card-id
+                         :dashboardId dashboard-id
+                         :databaseId  database-id
+                         :queryHash   (if (bytes? query-hash) (codecs/bytes->hex query-hash) nil)
```
This can be even further simplified:
```suggestion
                         :queryHash   (when (bytes? query-hash) (codecs/bytes->hex query-hash))
```

fernandobrito replied:
Thanks! I've updated the code with your suggestion.

Just for the sake of documentation: I was curious and checked what would happen if we remove the check for `bytes?` altogether. It causes for example table fingerprinting to fail with:

```
2023-11-26 14:28:34,024 WARN sync.util :: Error fingerprinting Table 81 '[redacted]'
java.lang.NullPointerException
        at org.apache.commons.codec.binary.Hex.encodeHex(Hex.java:171)
        at org.apache.commons.codec.binary.Hex.encodeHex(Hex.java:156)
        at org.apache.commons.codec.binary.Hex.encodeHex(Hex.java:142)
        at org.apache.commons.codec.binary.Hex.encodeHexString(Hex.java:290)
        ...
```

metamben commented:
@fernandobrito please let us know if you intend to change anything. I think the PR is good to go, but I would prefer to get a sign before merging.

fernandobrito commented:
> @fernandobrito please let us know if you intend to change anything. I think the PR is good to go, but I would prefer to get a sign before merging.

Hi! Thank you for writing a sample of how the function `query->native` could look like.  If possible, please wait until Monday or whenever you have time next week. I will have time this weekend to give it a final try and address your last comment. I also plan to do a final manual test and document on this PR how the query remark will look like in different scenarios (saved question, saved dashboard, subscription, alert, etc).

fernandobrito committed:
Merge branch 'master' into 28865-snowflake-remark

fernandobrito commented:
# Final manual tests

An unsaved question for example will trigger a query that looks like:

```
SELECT "PUBLIC"."categories"."id" AS "id", "PUBLIC"."categories"."name" AS "name" FROM "test-data"."PUBLIC"."categories" WHERE "PUBLIC"."categories"."id" = 2 LIMIT 2000

-- {"pulseId":null,"serverId":"9ef73cb3-8e76-4555-987f-92fa53bb85e6","client":"Metabase","queryHash":"e0fb6ae12c0e7c1afed355cf858910fba268b9926ade5d9408b349fa0581894d","actionId":null,"queryType":"query","cardId":null,"dashboardId":null,"context":"ad-hoc","userId":1,"databaseId":67}
```

The JSON prettified looks like:
```json
{
  "pulseId": null,
  "serverId": "9ef73cb3-8e76-4555-987f-92fa53bb85e6",
  "client": "Metabase",
  "queryHash": "e0fb6ae12c0e7c1afed355cf858910fba268b9926ade5d9408b349fa0581894d",
  "actionId": null,
  "queryType": "query",
  "cardId": null,
  "dashboardId": null,
  "context": "ad-hoc",
  "userId": 1,
  "databaseId": 67
}
```

I will perform a few more tests (triggering different scenarios) and I will paste below the query remark that was generated. For clarity, I will omit the SQL, prettify the JSON, remove the nulls, and also remove `serverId`, `client`, and `queryHash` since they are always present and I want to focus on the differences. I will also add an arrow to highlight any relevant fields.

### Unsaved question
Using Metabase's [query builder](https://www.metabase.com/glossary/query_builder).

```json
{
  "queryType": "query", <----
  "context": "ad-hoc",  <----
  "userId": 1,
  "databaseId": 67
}
```

Note how using the query builder, we get `"queryType": "query"` and `"context": "ad-hoc"`.

### Saved question
Using Metabase's [query builder](https://www.metabase.com/glossary/query_builder).

```json
{
  "queryType": "query",
  "cardId": 34, <----
  "context": "question", <----
  "userId": 1,
  "databaseId": 67
}
```

With the question saved, we now get a `cardId` and a different `context`.

### Unsaved question (manually written SQL)

```json
{
  "queryType": "native",  <----
  "context": "ad-hoc",
  "userId": 1,
  "databaseId": 67
}
```

If the question is created with your own SQL, `queryType` becomes `native`. I interpret that as `database native` SQL (in this case, Snowflake's native SQL).

### Viewing a saved question, through a dashboard

```json
{
  "queryType": "native",
  "cardId": 35,
  "dashboardId": 33,  <----
  "context": "dashboard",  <----
  "userId": 1,
  "databaseId": 67
}
``` 

In this scenario, we get both the `cardId` and the `dashboardId`. Note how the `context` was `dashboard`. 

### Viewing the question as a different user

In the query above, user `1` was both the author and the viewer of the question. What happens if I create a new user and trigger the question from it?

```json
{
  "queryType": "native",
  "cardId": 35,
  "dashboardId": 33,
  "context": "dashboard",
  "userId": 2, <----
  "databaseId": 67
}
```

The `userId` shows which user was logged in when the query was triggered, as expected.

### Editing a dashboard

I also realized that when you are editing a dashboard, you can trigger queries where the context will be `question` but you also get a `dashboardId`. For example, when you add a question to a dashboard.

```json
{
  "queryType": "native",
  "cardId": 35,
  "dashboardId": 33,  <----
  "context": "question",  <----
  "userId": 1,
  "databaseId": 67
}
```

fernandobrito committed:
Fix query remarks on Snowflake

fernandobrito committed:
Fix query remarks on Snowflake

fernandobrito commented:
@metamben your implementation suggestion worked well and now this PR is ready to be merged from my side. Thank you for your patience.

I also made 2 minor changes in the previous commit:
* I removed `actionId` from the query remark, as for now [Metabase actions](https://www.metabase.com/docs/latest/actions/introduction) are not supported on Snowflake, and this field in the remark would always be `null`. 
* In a previous commit (and between your reviews), I had refactored the `database-id` field to be taken from the `info` object ([here](https://github.com/metabase/metabase/pull/29888/commits/18aca90b250e74cbc2d49c89959168033315f9bc#diff-5472d9d00fa628881a6887a87e2be860f2d679f85aeada5d9f2c86173646800aL592)). Now that I did more extensive tests, I realized that this change made `databaseId` become `null`, so I reverted it to how it was before. 

All the Snowflake tests are passing locally, and I performed a few manual tests (results below).

fernandobrito commented:
# Final manual tests

An unsaved question for example will trigger a query that looks like:

```
SELECT "PUBLIC"."categories"."id" AS "id", "PUBLIC"."categories"."name" AS "name" FROM "test-data"."PUBLIC"."categories" WHERE "PUBLIC"."categories"."id" = 2 LIMIT 2000

-- {"pulseId":null,"serverId":"9ef73cb3-8e76-4555-987f-92fa53bb85e6","client":"Metabase","queryHash":"e0fb6ae12c0e7c1afed355cf858910fba268b9926ade5d9408b349fa0581894d","queryType":"query","cardId":null,"dashboardId":null,"context":"ad-hoc","userId":1,"databaseId":67}
```

The prettified JSON looks like:
```json
{
  "pulseId": null,
  "serverId": "9ef73cb3-8e76-4555-987f-92fa53bb85e6",
  "client": "Metabase",
  "queryHash": "e0fb6ae12c0e7c1afed355cf858910fba268b9926ade5d9408b349fa0581894d",
  "queryType": "query",
  "cardId": null,
  "dashboardId": null,
  "context": "ad-hoc",
  "userId": 1,
  "databaseId": 67
}
```

I will perform a few more tests (triggering queries through different scenarios) and I will paste below the query remark that was generated. For clarity, I will omit the SQL, prettify the JSON, remove the nulls, and also remove `serverId`, `client`, and `queryHash` since they are always present and I want to focus on the differences. I will also add an arrow to highlight any relevant fields.

### Unsaved question
Using Metabase's [query builder](https://www.metabase.com/glossary/query_builder).

```json
{
  "queryType": "query", <----
  "context": "ad-hoc",  <----
  "userId": 1,
  "databaseId": 67
}
```

Note how using the query builder, we get `"queryType": "query"` and `"context": "ad-hoc"`.

### Saved question
Using Metabase's [query builder](https://www.metabase.com/glossary/query_builder).

```json
{
  "queryType": "query",
  "cardId": 34, <----
  "context": "question", <----
  "userId": 1,
  "databaseId": 67
}
```

With the question saved, we now get a `cardId` and a different `context`.

### Unsaved question (manually written SQL)

```json
{
  "queryType": "native",  <----
  "context": "ad-hoc",
  "userId": 1,
  "databaseId": 67
}
```

If the question is created with your own SQL, `queryType` becomes `native`. I interpret that as `database native` SQL (in this case, Snowflake's native SQL).

### Viewing a saved question, through a dashboard

```json
{
  "queryType": "native",
  "cardId": 35,
  "dashboardId": 33,  <----
  "context": "dashboard",  <----
  "userId": 1,
  "databaseId": 67
}
``` 

In this scenario, we get both the `cardId` and the `dashboardId`. Note how the `context` was `dashboard`. 

### Viewing the question as a different user

In the query above, user `1` was both the author and the viewer of the question. What happens if I create a new user and trigger the question from it?

```json
{
  "queryType": "native",
  "cardId": 35,
  "dashboardId": 33,
  "context": "dashboard",
  "userId": 2, <----
  "databaseId": 67
}
```

The `userId` shows which user was logged in when the query was triggered, as expected. Internally, this property is even called `executed-by`.

### Editing a dashboard

I also realized that when you are editing a dashboard, you can trigger queries where the context will be `question` but you also get a `dashboardId`. For example, when you add a question to a dashboard.

```json
{
  "queryType": "native",
  "cardId": 35,
  "dashboardId": 33,  <----
  "context": "question",  <----
  "userId": 1,
  "databaseId": 67
}
```

### Testing a subscription
Testing it before it's created (by pressing the "send now" button).

```json
{
  "pulseId": null, <----
  "queryType": "native",
  "cardId": 35,
  "dashboardId": 33,
  "context": "pulse", <----
  "userId": 1,
  "databaseId": 67
}
```

In this case, since the subscription has not been created, we get `"context": "pulse"` but no `pulseId`, as expected.

### Scheduled subscription

```json
{
  "pulseId": null, <----
  "queryType": "native",
  "cardId": 35,
  "dashboardId": 33,
  "context": "pulse", <----
  "userId": 1,
  "databaseId": 67
}
```

⚠️  **Based on my tests, also the scheduled subscription had no `pulseId`, which is not what I expected.**

### Scheduled alerts

```json
{
  "pulseId": 2, <----
  "queryType": "query",
  "cardId": 37,
  "dashboardId": null, <----
  "context": "pulse", <----
  "userId": 1,
  "databaseId": 67
}
```

Interestingly enough, alerts (which internally are subscriptions/pulses) did receive the correct `pulseId` and the correct `context`. One way to differentiate alerts from subscriptions is that alerts are set on a question (while subscriptions are set on dashboards). As a consequence, `dashboardId` becomes `null` for alerts.

metamben commented:
> * I removed `actionId` from the query remark, as for now [Metabase actions](https://www.metabase.com/docs/latest/actions/introduction) are not supported on Snowflake, and this field in the remark would always be `null`.

Would it make sense to remove all properties with `nil` values in general?

fernandobrito commented:
For fun, here is an example of how to query and parse the query remarks from Snowflake's internal `QUERY_HISTORY`:

```sql
with source as (
    select 
        -- Parse the JSON object in the comment
        parse_json(regexp_substr(query_text, '-- ({.*})', 1, 1, 'c', 1)) as query_remark,
        *
        
    from snowflake.account_usage.query_history
    
    where
        -- Only rows with the query remark
        query_text like '%"client":"Metabase"%'

        -- This query itself would appear in the result, so we also filter by query remarks being parsed successfully
        and query_remark is not null
        
    order by start_time desc
    limit 100
)

select 
    query_remark:client::string as client,
    query_remark:serverId::string as server_id,
    query_remark:queryType::string as query_type,
    query_remark:userId::string as user_id,
    query_remark:context::string as context,
    query_remark:cardId::string as card_id,
    query_remark:dashboardId::string as dashboard_id,
    query_remark:databaseId::string as database_id,
    query_remark:pulseId::string as pulse_id,
    query_remark:queryHash::string as query_hash,
    
    * exclude (query_remark)
from source;
```

![image](https://github.com/metabase/metabase/assets/196840/8c4d6ede-8201-4af5-92a0-94002c3566eb)

metamben commented on modules/drivers/snowflake/test/metabase/driver/snowflake_test.clj:
```diff
@@ -298,3 +314,36 @@
+                                 "  OR (\"PUBLIC\".\"users\".\"id\" = 2)"
+                                 "  OR (\"PUBLIC\".\"users\".\"id\" = 3)"
+                                 "LIMIT"
+                                 "  2000 -- {\"pulseId\":null,\"serverId\":\"{{site-uuid}}\",\"client\":\"Metabase\",\"queryHash\":\"cb83d4f6eedc250edb0f2c16f8d9a21e5d42f322ccece1494c8ef3d634581fe2\",\"queryType\":\"query\",\"cardId\":1234,\"dashboardId\":5678,\"context\":\"ad-hoc\",\"userId\":1000,\"databaseId\":1}"]]
```
The tests are failing because `databaseId` is not usually `1`, you should get it as `(mt/id)`.

Also, testing the JSON representation is brittle, because Clojure maps are generally not guaranteed to keep the order. Small maps are implemented as `array-map`s which do keep the order, but after a certain size (above about 10 entries, so we are close) hash-maps are used and those don't.

So either use `array-map` in `snowflake.clj`, or (probably more complicated but better) parse the returned JSON string and compare Clojure maps.

fernandobrito replied:
Those are excellent points, thank you! I've pushed a new commit trying to address them (with the help from ChatGPT 😅).

By the way, what happens once we are done reviewing this PR? Will you merge this PR yourself? Or will you merge your mirror PR? I'm just asking so I know if I should do anything myself. In case I'm supposed to merge it myself, can I "squash & merge" since I have a lot of small commits?

metamben replied:
We squash merge your PR. The mirror PR is there just to run the tests.

fernandobrito committed:
Merge branch 'master' into 28865-snowflake-remark

fernandobrito committed:
Improve test on Snowflake query remark

fernandobrito committed:
Merge branch 'master' into 28865-snowflake-remark

fernandobrito committed:
Remove unnecessary comment

metamben commented on modules/drivers/snowflake/test/metabase/driver/snowflake_test.clj:
```diff
@@ -298,3 +312,35 @@
+                                      :query-hash   (byte-array [-53 -125 -44 -10 -18 -36 37 14 -37 15 44 22 -8 -39 -94 30
+                                                                 93 66 -13 34 -52 -20 -31 73 76 -114 -13 -42 52 88 31 -30])})))
+            ; re-find function returns a vector where the first element is the entire match and the subsequent elements
+            ; are the captured groups. So, (second ...) gives us the first captured group.
```
Thank you. When the mirror PR passes the tests, we merge this PR.

fernandobrito replied:
Thank @metamben for your support!
